Xamarin.Forms.Core/Shell/ShellNavigationManager.cs

```diff
@@ -0,0 +1,517 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Xamarin.Forms
+{
+	internal class ShellNavigationManager
+	{
+		readonly Shell _shell;
+		ShellNavigatedEventArgs _accumulatedEvent;
+		bool _accumulateNavigatedEvents;
+
+		public event EventHandler<ShellNavigatedEventArgs> Navigated;
+		public event EventHandler<ShellNavigatingEventArgs> Navigating;
+
+		public ShellNavigationManager(Shell shell)
+		{
+			_shell = shell;
+		}
+
+		public Task GoToAsync(ShellNavigationState state, bool? animate, bool enableRelativeShellRoutes, ShellNavigatingEventArgs deferredArgs = null)
+		{
+			return GoToAsync(new ShellNavigationParameters
+			{
+				TargetState = state,
+				Animated = animate,
+				EnableRelativeShellRoutes = enableRelativeShellRoutes,
+				DeferredArgs = deferredArgs
+			});
+		}
+
+		public async Task GoToAsync(ShellNavigationParameters shellNavigationParameters)
+		{
+			if (shellNavigationParameters.PagePushing != null)
+				Routing.RegisterImplicitPageRoute(shellNavigationParameters.PagePushing);
+
+			ShellNavigationState state = shellNavigationParameters.TargetState ?? Routing.GetRoute(shellNavigationParameters.PagePushing);
+			bool? animate = shellNavigationParameters.Animated;
+			bool enableRelativeShellRoutes = shellNavigationParameters.EnableRelativeShellRoutes;
+			ShellNavigatingEventArgs deferredArgs = shellNavigationParameters.DeferredArgs;
+
+			var navigationRequest = ShellUriHandler.GetNavigationRequest(_shell, state.FullLocation, enableRelativeShellRoutes, shellNavigationParameters: shellNavigationParameters);
+			bool isRelativePopping = ShellUriHandler.IsTargetRelativePop(shellNavigationParameters);
+
+			ShellNavigationSource source = CalculateNavigationSource(_shell, _shell.CurrentState, navigationRequest);
+
+			// If the deferredArgs are non null that means we are processing a delayed navigation
+			// so the user has indicated they want to go forward with navigation
+			// This scenario only comes up from UI iniated navigation (i.e. switching tabs)
+			if (deferredArgs == null)
+			{
+				var navigatingArgs = ProposeNavigation(source, state, _shell.CurrentState != null);
+
+				bool accept = !navigatingArgs.NavigationDelayedOrCancelled;
+				if (navigatingArgs.DeferredTask != null)
+					accept = await navigatingArgs.DeferredTask;
```

Comment:
Before this PR a deferred GotoAsync would always call `GotoAsync` again once the operation was completed but with these changes the same `GotoAsync` just continues on after the user has completed the deferral